File: src/router/libyang/swig/cpp/examples/xpath.cpp
/**
 * @file xpath.cpp
 * @author Mislav Novakovic <mislav.novakovic@sartura.hr>
 * @brief Example of the libyang C++ bindings
 *
 * Copyright (c) 2017 Deutsche Telekom AG.
 *
 * This source code is licensed under BSD 3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://opensource.org/licenses/BSD-3-Clause
 */

#include <iostream>
#include <memory>
#include <stdexcept>
#include <vector>

#include <Libyang.hpp>
#include <Tree_Data.hpp>
#include <Tree_Schema.hpp>

int main() {

    libyang::S_Context ctx;
    try {
        ctx = std::make_shared<libyang::Context>("/etc/sysrepo/yang");
        auto module = ctx->load_module("turing-machine", nullptr);

        auto node = ctx->parse_data_path("/etc/sysrepo/data/turing-machine.startup", LYD_XML, LYD_OPT_CONFIG);

        auto node_set = node->find_path("/turing-machine:turing-machine/transition--function/delta[label='left summand']/*");\
        if (!node_set) {
            std::cout << "could not find data for xpath" << std::endl;
            return -1;
        }

        auto list = node_set->data();
        for(auto data_set = list.begin() ; data_set != list.end() ; ++data_set) {
            std::cout << "name: " << (*data_set)->schema()->name() << " type: " << (*data_set)->schema()->nodetype() << " path: " << (*data_set)->path() << std::endl;
        }
    } catch( const std::exception& e ) {
        std::cout << "test" << std::endl;
        std::cout << e.what() << std::endl;
        auto errors = libyang::get_ly_errors(ctx);
        for(auto error = errors.begin() ; error != errors.end() ; ++error) {
            std::cout << "err: " << (*error)->err() << std::endl;
            std::cout << "vecode: " << (*error)->vecode() << std::endl;
            std::cout << "errmsg: " << (*error)->errmsg() << std::endl;
            std::cout << "errpath: " << (*error)->errpath() << std::endl;
            std::cout << "errapptag: " << (*error)->errapptag() << std::endl;
        }
        return -1;
    }

    return 0;
}

File: src/router/libyang/swig/python/examples/xpath.py
__author__ = "Mislav Novakovic <mislav.novakovic@sartura.hr>"
__copyright__ = "Copyright 2017, Deutsche Telekom AG"
__license__ = "BSD 3-Clause"

import yang as ly
import sys

ctx = None
try:
    ctx = ly.Context("/etc/sysrepo/yang")

    module = ctx.load_module("turing-machine", None)
    if module is None:
        print("module not loaded")
        sys.exit()

    node = None
    try:
        if node is None : node = ctx.parse_data_path("/etc/sysrepo/data/turing-machine.startup", ly.LYD_LYB, ly.LYD_OPT_CONFIG)
    except Exception as e:
        print(e)
    try:
        if node is None : node = ctx.parse_data_path("/etc/sysrepo/data/turing-machine.startup", ly.LYD_XML, ly.LYD_OPT_CONFIG)
    except Exception as e:
        print(e)
    try:
        if node is None : node = ctx.parse_data_path("/etc/sysrepo/data/turing-machine.startup", ly.LYD_JSON, ly.LYD_OPT_CONFIG)
    except Exception as e:
        print(e)

    if node is None:
        sys.exit()

    node_set = node.find_path("/turing-machine:turing-machine/transition-function/delta[label='left summand']/*")
    if node_set is None:
        print("could not find data for xpath")
        sys.exit()

    for data_set in node_set.data():
        schema = data_set.schema()
        print("name: " + schema.name() + " type: " + str(schema.nodetype()) + " path: " + data_set.path())

except Exception as e:
    print(e)
    errors = ly.get_ly_errors(ctx)
    for err in errors:
        print("err: %d" % err.err())
        print("vecode: %d" % err.vecode())
        print("errmsg: " + err.errmsg())
        print("errpath:" + err.errpath())
        print("errapptag:" + err.errapptag())
